apps/internal/base/base.go

```diff
@@ -302,3 +306,30 @@ func convertStrUnixToUTCTime(unixTime string) (time.Time, error) {
 	}
 	return time.Unix(timeInt, 0).UTC(), nil
 }
+
+func getKeyFromRequest(authParams authority.AuthParams) string {
+	if authParams.AuthorizationType == authority.ATClientCredentials {
+		return getAppKey(authParams)
+	}
+	if authParams.AuthorizationType == authority.ATRefreshToken {
+		return authParams.HomeaccountID
+	}
+	return ""
+}
+
+func getKeyFromResponse(authParams authority.AuthParams, homeaccountIDfromResponse string) string {
+	if authParams.AuthorizationType == authority.ATClientCredentials {
+		return getAppKey(authParams)
+	}
+	if authParams.IsConfidentialClient || authParams.AuthorizationType == authority.ATRefreshToken {
+		return homeaccountIDfromResponse
+	}
+	return ""
+}
+
+func getAppKey(authParams authority.AuthParams) string {
```

Comment:
Same, AppKey() and attach to authority.AuthParams.

Reply:
Resoved in 71dea81